`<variant>`: Current copy assignment operator's implementation rejects types that has explicit copy constructor

# Describe the bug
According to the C++23 standard (N4950), the copy assignment operator of `std::variant` is specified to create a temporary object by copy-constructing a `std::variant` and to use the move assignment operator when the selected alternative type of the right-hand side operand has a potentially throwing copy constructor and a non-throwing move constructor (see [\[variant.assign\]\#2.5](https://timsong-cpp.github.io/cppwp/n4950/variant.assign#2.5)). In such a case, the contained value of the temporary object is required to be **direct-initialized**.

However, [the current implementation](https://github.com/microsoft/STL/blob/e2ef398685f7e470dbaeaf65ff919de72bda7489/stl/inc/variant#L717) performs **copy-initialization** of the temporary object instead. Consequently, certain valid types—such as those with an explicit copy constructor—are rejected by the compiler, which is non-conforming behavior.

# Command-line test case
```
C:\Temp>type repro.cpp
#include <variant>

struct S {
  S() noexcept = default;

  explicit S(S const&) noexcept(false) = default;  // potentially throwing and explicit copy constructor
  S(S&&) noexcept = default;  // no throwing move constructor

  S& operator=(S const&) noexcept { return *this; }  // non-trivial copy assignment operator
};

int main()
{
  std::variant<int, S> a = 12;
  std::variant<int, S> b = S{};
  a = b;
}


C:\Temp>cl /EHsc /W4 /WX /std:c++latest .\repro.cpp
Microsoft(R) C/C++ Optimizing Compiler Version 19.50.35724 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest が最新の C++ Working Draft の言語機能のプレビューとして提供されています。
バグの報告や機能改善の提案を歓迎いたします。
ただし、これらの機能は現状のままサポートなしで提供されており、
Working Draft の進展に伴い変更や削除が生じることにご注意ください。詳細については、
https://go.microsoft.com/fwlink/?linkid=2045807 をご覧ください。

Souce.cpp
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(711): error C2440: '初期化中': 'const _First' から 'S' に変換できません。
        with
        [
            _First=S
        ]
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(711): note: struct 'S' のコンストラクターが 'explicit' と宣言されています。
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(711): note: テンプレートインスタンス化コンテキスト (最初に最も古いもの) は
Souce.cpp(14): note: コンパイル対象の クラス テンプレート インスタンス化 'std::variant<int,S>' のリファレンスを確認してください
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(931): note: コンパイル対象の クラス テンプレート インスタンス化 'std::_Non_trivial_move_assign<std::_Variant_base<int,S>,int,S>' のリファレンスを確認してください
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\xsmf_control.h(122): note: コンパイル対象の クラス テンプレート インスタンス化 'std::_Non_trivial_copy_assign<std::_Variant_base<int,S>,int,S>' のリファレンスを確認してください
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\xsmf_control.h(85): note: コンパイル対象の クラス テンプレート インスタンス化 'std::_Variant_base<int,S>' のリファレンスを確認してください
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(827): note: クラス テンプレート メンバー関数 'void std::_Variant_base<int,S>::_Assign_from(const std::_Variant_base<int,S> &) noexcept(false)' のコンパイル中
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\xsmf_control.h(95): note: 'std::_Non_trivial_copy_assign<std::_Variant_base<int,S>,int,S>::operator =' の 'std::_Variant_base<int,S>::_Assign_from' への最初の参照を参照してください
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\xsmf_control.h(129): note: 'std::_Non_trivial_move_assign<std::_Variant_base<int,S>,int,S>::operator =' の 'std::_Non_trivial_copy_assign<std::_Variant_base<int,S>,int,S>::operator =' への最初の参照を参照してください
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(830): note: コンパイル対象の関数 テンプレート インスタンス化 'void std::_Variant_raw_visit<const std::_Variant_storage_<true,int,S>&,std::_Variant_assign_visitor<int,S>>(size_t,_Storage,_Fn &&) noexcept(false)' のリファレンスを確認してください
        with
        [
            _Storage=const std::_Variant_storage_<true,int,S> &,
            _Fn=std::_Variant_assign_visitor<int,S>
        ]
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(653): note: コンパイル対象の関数 テンプレート インスタンス化 'void std::_Variant_raw_visit1<1>::_Visit<std::_Variant_assign_visitor<int,S>,const std::_Variant_storage_<true,int,S>&>(size_t,_Fn &&,_Storage) noexcept(false)' のリファレンスを確認してください
        with
        [
            _Fn=std::_Variant_assign_visitor<int,S>,
            _Storage=const std::_Variant_storage_<true,int,S> &
        ]
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\variant(608): note: コンパイル対象の関数 テンプレート インスタンス化 'void std::_Variant_assign_visitor<int,S>::operator ()<const _First&,1>(std::_Tagged<const _First &,1>) noexcept(false) const' のリファレンスを確認してください
        with
        [
            _First=S
        ]
```

# Expected behavior


# STL version
```
C:\Temp>type meow.cpp
#include <iostream>
int main() {
    std::cout << _MSVC_STL_UPDATE << "\n";
}

C:\Temp>cl /EHsc /nologo /W4 meow.cpp && meow
meow.cpp
202508
```

# Additional context
[libstdc++'s implementation](https://github.com/gcc-mirror/gcc/blob/3a1bb1b70d017e4816650ed42d316accf2d93944/libstdc%2B%2B-v3/include/std/variant#L695) and [libc++'s implementation](https://github.com/llvm/llvm-project/blob/945db339632895e18721882b0a71e9d41d9c4ef7/libcxx/include/variant#L923) both perform **direct-initialization**.

related LWG issue: https://cplusplus.github.io/LWG/issue2904

Sorry about the language of the compiler output is Japanese. I don't know how to change it.

--- a/stl/inc/variant
+++ b/stl/inc/variant
@@ -714,7 +714,7 @@ struct _Variant_assign_visitor { // visitor that implements assignment for varia
                         _Self._Reset();
                         _STD _Construct_in_place(_Self._Storage(), integral_constant<size_t, _Idx>{}, _Source._Val);
                     } else { // copy throws and move does not; move from a temporary copy
-                        auto _Temp = _Source._Val;
+                        auto _Temp(_Source._Val);
                         _Self._Reset();
                         _STD _Construct_in_place(_Self._Storage(), integral_constant<size_t, _Idx>{}, _STD move(_Temp));
                     }

--- a/tests/std/tests/P0088R3_variant_msvc/test.cpp
+++ b/tests/std/tests/P0088R3_variant_msvc/test.cpp
@@ -22,6 +22,11 @@
 #include <type_id.h>
 #include <variant_test_helpers.h>
 
+#if _HAS_CXX20
+#define CONSTEXPR20 constexpr
+#else // ^^^ _HAS_CXX20 / !_HAS_CXX20 vvv
+#define CONSTEXPR20 inline
+#endif // ^^^ !_HAS_CXX20 ^^^
 namespace msvc {
     namespace size {
         template <class T>
@@ -743,11 +748,6 @@ namespace msvc {
     } // namespace gh2770
 
     namespace gh4901 {
-#if _HAS_CXX20
-#define CONSTEXPR20 constexpr
-#else // ^^^ _HAS_CXX20 / !_HAS_CXX20 vvv
-#define CONSTEXPR20 inline
-#endif // ^^^ !_HAS_CXX20 ^^^
         struct X {
             CONSTEXPR20 ~X() {}
         };
@@ -822,7 +822,6 @@ namespace msvc {
 
         static_assert(std::is_nothrow_destructible_v<std::variant<Y, int, Y2>>);
         static_assert(std::is_nothrow_destructible_v<ZC>);
-#undef CONSTEXPR20
     } // namespace gh4901
 
     namespace assign_cv {
@@ -937,9 +936,40 @@ namespace msvc {
             std::variant<std::optional<GenericSpec>> u;
         };
     } // namespace gh4959
+
+    namespace gh6085 {
+        struct WeirdlyCopyConstructible {
+            WeirdlyCopyConstructible() = default;
+            // potentially-throwing and explicit
+            explicit WeirdlyCopyConstructible(const WeirdlyCopyConstructible&) noexcept(false) = default;
+            WeirdlyCopyConstructible(WeirdlyCopyConstructible&&) noexcept                      = default;
+
+            CONSTEXPR20 WeirdlyCopyConstructible& operator=(const WeirdlyCopyConstructible&) noexcept { // non-trivial
+                return *this;
+            }
+        };
+
+        CONSTEXPR20 bool run_test() {
+            std::variant<int, WeirdlyCopyConstructible> a = 12;
+            std::variant<int, WeirdlyCopyConstructible> b = WeirdlyCopyConstructible{};
+
+            assert(a.index() == 0);
+            assert(std::get<0>(a) == 12);
+            assert(b.index() == 1);
+
+            a = b;
+            assert(a.index() == 1);
+
+            return true;
+        }
+    } // namespace gh6085
 } // namespace msvc
 
 int main() {
+#if _HAS_CXX20
+    static_assert(msvc::gh6085::run_test());
+#endif // _HAS_CXX20
+
     msvc::big_variant::run_test();
     msvc::derived_variant::run_test();
     msvc::visit::run_test();
@@ -951,5 +981,6 @@ int main() {
     msvc::vso492097::run_test();
     msvc::DevCom1031281::run_test();
     msvc::gh2770::run_test();
+    msvc::gh6085::run_test();
     msvc::assign_cv::run_test();
 }

